src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java

```diff
@@ -970,6 +979,407 @@ public void recursivelyDelete(final Project project, final boolean commitIndex)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter, queryParameter);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """                        
+                    DELETE FROM "FINDINGATTRIBUTION" WHERE "COMPONENT_ID" IN (
+                        SELECT "ID" FROM "COMPONENT" WHERE "PROJECT_ID" IN (
+                            SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                        )
+                    ) AND "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """.replaceAll(Pattern.quote("= ANY(?)"), inExpression)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter, queryParameter);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "COMPONENTS_VULNERABILITIES" WHERE "COMPONENT_ID" IN (
+                        SELECT "ID" FROM "COMPONENT" WHERE "PROJECT_ID" IN (
+                            SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                        )
+                    );
+                    """.replace("= ANY(?)", inExpression)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """
+                    DELETE FROM "ANALYSISCOMMENT" WHERE "ANALYSIS_ID" IN (
+                        SELECT "ID" FROM "ANALYSIS" WHERE "PROJECT_ID" IN (
+                            SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                        )
+                    );
+                    """.replace("= ANY(?)", inExpression)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """                        
+                    DELETE FROM "ANALYSIS" WHERE "PROJECT_ID" IN (
+                        SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                    );
+                    """.replace("= ANY(?)", inExpression)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter);
+
+                sqlQuery = pm.newQuery(JDOQuery.SQL_QUERY_LANGUAGE, """                        
+                    DELETE FROM "COMPONENT_PROPERTY" WHERE "COMPONENT_ID" IN (
+                        SELECT "ID" FROM "COMPONENT" WHERE "PROJECT_ID" IN (
+                            SELECT "ID" FROM "PROJECT" WHERE "UUID" = ANY(?)
+                        )
+                    );
+                    """.replace("= ANY(?)", inExpression)
+                );
+                executeAndCloseWithArray(sqlQuery, queryParameter);
+
+                // Deletion with CTEs does not work with H2, but verified on Postgres and MS SQL Server
+                if ( ! Config.getInstance().getProperty(Config.AlpineKey.DATABASE_DRIVER).equals("org.h2.Driver")) {
```

Comment:
You can use `DbUtil.isH2()`, `DbUtil.isMssql()` etc. to make these conditions less verbose.